backend/api/v1/project_service.go

```diff
@@ -360,72 +360,90 @@ func (s *ProjectService) BatchGetIamPolicy(ctx context.Context, request *v1pb.Ba
-	if err != nil {
-		return nil, status.Errorf(codes.Internal, "failed to convert to roles: %v", err)
-	}
-	if err := s.validateIAMPolicy(ctx, request.Policy, roles); err != nil {
-		return nil, status.Errorf(codes.InvalidArgument, err.Error())
-	}
-	project, err := s.store.GetProjectV2(ctx, &store.FindProjectMessage{
-		ResourceID: &projectID,
-	})
-	if err != nil {
-		return nil, status.Errorf(codes.Internal, err.Error())
-	}
-	if project == nil {
-		return nil, status.Errorf(codes.NotFound, "project %q not found", request.Resource)
-	}
-	if project.Deleted {
-		return nil, status.Errorf(codes.NotFound, "project %q has been deleted", request.Resource)
-	}
+	var oriIamPolicyMsg *store.IamPolicyMessage
+	var newIamPolicyMsg *store.IamPolicyMessage
+	var deltas []*v1pb.BindingDelta
 
-	policy, err := convertToStoreIamPolicy(ctx, s.store, request.Policy)
-	if err != nil {
-		return nil, status.Errorf(codes.Internal, err.Error())
-	}
+	v1pbIamPolicy, apiErr := func() (*v1pb.IamPolicy, error) {
+		projectID, err := common.GetProjectID(request.Resource)
+		if err != nil {
+			return nil, status.Errorf(codes.InvalidArgument, err.Error())
+		}
+		creatorUID, ok := ctx.Value(common.PrincipalIDContextKey).(int)
+		if !ok {
+			return nil, status.Errorf(codes.Internal, "cannot get principal ID from context")
+		}
+		roleMessages, err := s.store.ListRoles(ctx)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "failed to list roles: %v", err)
+		}
+		roles, err := convertToRoles(ctx, s.iamManager, roleMessages)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "failed to convert to roles: %v", err)
+		}
+		if err := s.validateIAMPolicy(ctx, request.Policy, roles); err != nil {
+			return nil, status.Errorf(codes.InvalidArgument, err.Error())
+		}
+		project, err := s.store.GetProjectV2(ctx, &store.FindProjectMessage{
+			ResourceID: &projectID,
+		})
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, err.Error())
+		}
+		if project == nil {
+			return nil, status.Errorf(codes.NotFound, "project %q not found", request.Resource)
+		}
+		if project.Deleted {
+			return nil, status.Errorf(codes.NotFound, "project %q has been deleted", request.Resource)
+		}
 
-	policyMessage, err := s.store.GetProjectIamPolicy(ctx, project.UID)
-	if err != nil {
-		return nil, status.Errorf(codes.Internal, "failed to find project iam policy with error: %v", err.Error())
-	}
-	if request.Etag != policyMessage.Etag {
-		return nil, status.Errorf(codes.Aborted, "there is concurrent update to the project iam policy, please refresh and try again.")
-	}
+		policy, err := convertToStoreIamPolicy(ctx, s.store, request.Policy)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, err.Error())
+		}
 
-	policyPayload, err := protojson.Marshal(policy)
-	if err != nil {
-		return nil, status.Errorf(codes.Internal, err.Error())
+		policyMessage, err := s.store.GetProjectIamPolicy(ctx, project.UID)
```

Comment:
Let's minimize the change here. All we need to do is to get the old policy before the policy update. Then persist the diff of old/new policy.